tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/contact/EmailAddressContact.scala

```diff
@@ -29,21 +34,40 @@ case class AccountEmailContact(accountId: String, id: UUID, address: MailAddress
 trait EmailAddressContactSearchEngine {
   def index(accountId: AccountId, address: MailAddress): Publisher[EmailAddressContact]
 
+  def index(accountId: AccountId, address: MailAddress, firstname: String, surname: String): Publisher[EmailAddressContact]
+
+  def index(domain: Domain, address: MailAddress, firstname: String, surname: String): Publisher[EmailAddressContact]
+
   def autoComplete(accountId: AccountId, part: String): Publisher[EmailAddressContact]
 
 }
 
 class InMemoryEmailAddressContactSearchEngine extends EmailAddressContactSearchEngine {
   val emailList: Multimap[AccountId, EmailAddressContact] = Multimaps.synchronizedSetMultimap(HashMultimap.create())
+  val domainList: Multimap[Domain, EmailAddressContact] = Multimaps.synchronizedSetMultimap(HashMultimap.create())
 
   override def index(accountId: AccountId, address: MailAddress): Publisher[EmailAddressContact] =
     index(accountId, EmailAddressContact.of(address))
 
+  override def index(accountId: AccountId, address: MailAddress, firstname: String, surname: String): Publisher[EmailAddressContact] =
+    index(accountId, EmailAddressContact.of(address, firstname, surname))
+
   private def index(accountId: AccountId, addressContact: EmailAddressContact) =
     SMono.fromCallable(() => emailList.put(accountId, addressContact))
       .`then`(SMono.just(addressContact))
 
-  override def autoComplete(accountId: AccountId, part: String): Publisher[EmailAddressContact] =
+  override def index(domain: Domain, address: MailAddress, firstname: String, surname: String): Publisher[EmailAddressContact] =
+    index(domain, EmailAddressContact.of(address, firstname, surname))
+
+  private def index(domain: Domain, addressContact: EmailAddressContact): Publisher[EmailAddressContact] =
+    SMono.fromCallable(() => domainList.put(domain, addressContact))
+      .`then`(SMono.just(addressContact))
+
+  override def autoComplete(accountId: AccountId, part: String): Publisher[EmailAddressContact] = {
+    val domain = Username.of(accountId.getIdentifier).getDomainPart.get // TODO: evaluate optional
```

Comment:
Will work on it, the mix Scala, reactor, optional was a pain :)

Reply:
Yes here we loose the ability to do JMAP without virtual hosting...

Let me suggest:

```
    val maybeDomain: Option[Domain] = Username.of(accountId.getIdentifier).getDomainPart.asScala // needs import OptionalUtils for scala-jdk8 compat

    SFlux.concat(
          maybeDomain..map(domain => SFlux.fromIterable(domainList.get(domain).asScala).getOrElse(SFlux.empty()),
          SFlux.fromIterable(emailList.get(accountId).asScala))
      .filter(_.contains(part))
```